proxy/src/dns.rs

```diff
@@ -1,68 +1,61 @@
-use abstract_ns;
-use abstract_ns::HostResolve;
-pub use abstract_ns::IpList;
-use domain;
 use futures::prelude::*;
-use ns_dns_tokio;
 use std::fmt;
 use std::net::IpAddr;
-use std::path::Path;
-use std::str::FromStr;
 use std::time::Duration;
 use tokio_core::reactor::{Handle, Timeout};
 use transport;
+use trust_dns_resolver;
+use trust_dns_resolver::config::{ResolverConfig, ResolverOpts};
+use trust_dns_resolver::error::{ResolveError, ResolveErrorKind};
+use trust_dns_resolver::ResolverFuture;
+use trust_dns_resolver::lookup_ip::LookupIpFuture;
+use trust_dns_resolver::lookup_ip::LookupIp;
 
 #[derive(Clone, Debug)]
-pub struct Config(domain::resolv::ResolvConf);
+pub struct Config {
+    config: ResolverConfig,
+    opts: ResolverOpts,
+}
 
 #[derive(Clone, Debug)]
 pub struct Resolver {
-    resolver: ns_dns_tokio::DnsResolver,
+    config: Config,
     executor: Handle,
 }
 
 pub enum IpAddrFuture {
-    DNS(ns_dns_tokio::HostFuture),
+    DNS(LookupIpFuture),
     Fixed(IpAddr),
 }
 
 pub enum Error {
     NoAddressesFound,
-    ResolutionFailed(<ns_dns_tokio::HostFuture as Future>::Error),
+    ResolutionFailed(ResolveError),
 }
 
 pub enum Response {
-    Exists(IpList),
+    Exists(LookupIp),
     DoesNotExist,
 }
 
 // `Box<Future>` implements `Future` so it doesn't need to be implemented manually.
-pub type IpAddrListFuture = Box<Future<Item=Response, Error=abstract_ns::Error>>;
+pub type IpAddrListFuture = Box<Future<Item=Response, Error=ResolveError>>;
 
 /// A DNS name.
 #[derive(Clone, Debug, Eq, Hash, PartialEq)]
-pub struct Name(abstract_ns::Name);
+pub struct Name(String);
 
 impl fmt::Display for Name {
     fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
         self.0.fmt(f)
     }
 }
 
-impl Name {
-    /// Parses the input string as a DNS name, normalizing it to lowercase.
-    pub fn normalize(s: &str) -> Result<Self, ()> {
-        // XXX: `abstract_ns::Name::from_str()` wrongly accepts IP addresses as
-        // domain names. Protect against this. TODO: Fix abstract_ns.
-        if let Ok(_) = IpAddr::from_str(s) {
-            return Err(());
-        }
-        // XXX: `abstract_ns::Name::from_str()` doesn't accept uppercase letters.
-        //  TODO: Avoid this extra allocation.
-        let s = s.to_ascii_lowercase();
-        abstract_ns::Name::from_str(&s)
-            .map(Name)
-            .map_err(|_| ())
+impl<'a> From<&'a str> for Name {
+    fn from(s: &str) -> Self {
+        // TODO: Verify the name is a valid DNS name.
+        // TODO: Avoid this extra allocation.
+        Name(s.to_ascii_lowercase())
```

Comment:
My understanding is the http create will validate these names for us before we extract them. Our use of `abstract_ns::Name::from_str()` for validation was misguided (based on discussion with the maintainer of abstract_ns) and didn't catch all potential issues anyway.